File: 435. Non-overlapping Intervals/greedy.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
	int eraseOverlapIntervals(vector<Interval>& intervals)
	{
		if (intervals.size() == 0) { return (0); }
		sort(intervals.begin(), intervals.end(), mysort);

		int end = intervals[0].end;
		int erased = 0;
		for (int i = 1; i < intervals.size(); i++)
		{
			if (intervals[i].start >= end) {
				end = intervals[i].end;
			} else {
				erased++;
			}
		}
		return (erased);
	}


	static bool mysort(Interval& a, Interval& b)
	{
		return (a.end < b.end);
	}
}

File: 435. Non-overlapping Intervals/greedy.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e


class Solution(object):
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        end = float('-inf')
        erased = 0
        for i in sorted(intervals, key=lambda x: x.end):
            if i.start >= end:
                end = i.end
            else:
                erased += 1
        return erased
